### PR TITLE
Add data prepper documentation for grok performance_metadata

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
Adds documentation for grok `performance_metadata` parameter

### Issues Resolved
Related to https://github.com/opensearch-project/data-prepper/pull/4230

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by graytaylor0** [comment]
Not sure why DCO is failing. This commit was done with `git commit -s` , and the commit in my repo shows the signoff

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 1]
```diff
@@ -15,22 +15,20 @@ The Grok processor uses pattern matching to structure and extract important keys
...
-pattern_files_glob | No | String | Specifies which pattern files to use from the directories specified for `pattern_directories`. Default value is `*`.
-target_key | No | String | Specifies a parent-level key used to store all captures. Default value is `null`.
-timeout_millis | No | Integer | The maximum amount of time during which matching occurs. Setting to `0` disables the timeout. Default value is `30,000`.
-
-<!---## Configuration
-
-Content will be added to this section.--->
+:--- | :--- |:--- | :---
+`break_on_match` | No | Boolean | Specifies whether to match all patterns (`true`) or stop once the first successful match is found (`false`). Defaultis `true`.
+`grok_when` | No | String  | Specifies under what condition the `Grok` processor should perform matching. Default is no condition.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:20]:
```suggestion
`grok_when` | No | String  | Specifies under what condition the `grok` processor should perform matching. Default is no condition.
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 2]
```diff
@@ -15,22 +15,20 @@ The Grok processor uses pattern matching to structure and extract important keys
...
+`match` | No | Map | Specifies which keys to match specific patterns against. Default is an empty response body.
+`named_captures_only` | No | Boolean | Specifies whether to keep only named captures. Default is `true`.
+`pattern_definitions` | No | Map | Allows for custom pattern which can be used inline. Dafault is an empty response body.
+`patterns_directories` | No | List | Specifies directory paths that contain the customer pattern files. Default is an empty list.
+`pattern_files_glob` | No | String | Specifies which pattern files to use from the directories specified for `pattern_directories`. Default is `*`.
+`target_key` | No | String | Specifies a parent-level key used to store all captures. Default value is `null`.
+`timeout_millis` | No | Integer | The maximum amount of time during which matching occurs. Setting to `0` disables the timeout. Default is `30,000`.
+`performance_metadata` | No | Boolean | Whether or not to add performance metadata to Events. Defaults to `false`. For more information, see [Grok Performance Metadata](#grok-performance-metadata)
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:30]:
```suggestion
`performance_metadata` | No | Boolean | Whether or not to add performance metadata to events. Default is `false`. For more information, see [Grok Performance Metadata](#grok-performance-metadata)
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 3]
```diff
@@ -48,6 +46,33 @@ processor:
 ```
 The `grok_when` option can take a conditional expression. This expression is detailed in the [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/) documentation.
 
+# Grok performance metadata
+
+When `performance_metadata` is set to true, the `grok` processor adds the following metadata keys to each event.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:51]:
```suggestion
When the `performance_metadata` option is set to true, the `grok` processor adds the following metadata keys to each event:
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 4]
```diff
@@ -48,6 +46,33 @@ processor:
...
+
+* `_total_grok_processing_time` - The total time, in milliseconds, that the `grok` processor takes to match the event. This is the sum processing time for this event based on all of the grok processors that ran on this Event, and also have the `performance_metadata` enabled.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:53]:
```suggestion
* `_total_grok_processing_time` - The total time, in milliseconds, that the `grok` processor takes to match the event. This is the sum of the processing time for this event based on all of the grok processors that ran on this event that have the `performance_metadata` option enabled.
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 5]
```diff
@@ -48,6 +46,33 @@ processor:
...
+* `_total_grok_patterns_attempted` - The total number `grok` pattern match attempts across all `grok` processors that have run on the event.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:54]:
```suggestion
* `_total_grok_patterns_attempted` - The total number of `grok` pattern match attempts across all `grok` processors that have run on the event.
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 6]
```diff
@@ -48,6 +46,33 @@ processor:
...
+
+To include this metadata in the event that is sent to the sink of the pipeline, and `add_entries` processor can be used.
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:56]:
```suggestion
To include this metadata in the event when the event is sent to the sink set inside the pipeline, use the `add_entries` processor tot describe the metadata you want to include, as shown in the following example:
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 7]
```diff
@@ -48,6 +46,33 @@ processor:
...
+
+If you want to include grok performance metadata when the event is sent from the processor to the sink, use the `add_entries` processor to describe the metadata, as shown in the following example:
```
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:58]:
```suggestion

```

**Review by Naarcha-AWS — APPROVED** [review]

**Comment by Naarcha-AWS** [comment]
@dlvenable: Let me know if this is good to merge.

**Review by dlvenable — APPROVED** [review]

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 8]
```diff
@@ -15,26 +15,25 @@ The Grok processor uses pattern matching to structure and extract important keys
...
-
-<!---## Configuration
-
-Content will be added to this section.--->
+:--- | :--- |:--- | :---
+`break_on_match` | No | Boolean | Specifies whether to match all patterns (`true`) or stop once the first successful match is found (`false`). Default is `true`.
+`grok_when` | No | String  | Specifies under what condition the `grok` processor should perform matching. Default is no condition.
+`keep_empty_captures` | No | Boolean | Enables the preservation of `null` captures from the processed output. Default is `false`.
+`keys_to_overwrite` | No | List | Specifies which existing keys will be overwritten if there is a capture with the same key value. Default is `[]`.
+`match` | No | Map | Specifies which keys to should match specific patterns. Default is an empty response body.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:23]:
```suggestion
`match` | No | Map | Specifies which keys should match specific patterns. Default is an empty response body.
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 9]
```diff
@@ -15,26 +15,25 @@ The Grok processor uses pattern matching to structure and extract important keys
...
+`named_captures_only` | No | Boolean | Specifies whether to keep only named captures. Default is `true`.
+`pattern_definitions` | No | Map | Allows fors custom pattern that can be used inline inside the response body. Default is an empty response body.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:25]:
```suggestion
`pattern_definitions` | No | Map | Allows for a custom pattern that can be used inline inside the response body. Default is an empty response body.
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 10]
```diff
@@ -15,26 +15,25 @@ The Grok processor uses pattern matching to structure and extract important keys
...
+`patterns_directories` | No | List | Specifies which directory paths contain the custom pattern files. Default is an empty list.
+`pattern_files_glob` | No | String | Specifies which pattern files to use from the directories specified for `pattern_directories`. Default is `*`.
+`target_key` | No | String | Specifies a parent-level key used to store all captures. Default value is `null`.
+`timeout_millis` | No | Integer | The maximum amount of time during which matching occurs. Setting to `0` prevents any matching from occuring. Default is `30,000`.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:29]:
```suggestion
`timeout_millis` | No | Integer | The maximum amount of time during which matching occurs. Setting to `0` prevents any matching from occurring. Default is `30,000`.
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 11]
```diff
@@ -15,26 +15,25 @@ The Grok processor uses pattern matching to structure and extract important keys
...
+`performance_metadata` | No | Boolean | Whether or not to add the performance metadata to events. Default is `false`. For more information, see [Grok performance metadata](#grok-performance-metadata)
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:30]:
```suggestion
`performance_metadata` | No | Boolean | Whether or not to add the performance metadata to events. Default is `false`. For more information, see [Grok performance metadata](#grok-performance-metadata).
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 12]
```diff
@@ -46,8 +45,36 @@ processor:
         match:
           message: ['%{IPV6:clientip} %{WORD:request} %{POSINT:bytes}']
 ```
+{% include copy.html %}
+
 The `grok_when` option can take a conditional expression. This expression is detailed in the [Expression syntax](https://opensearch.org/docs/latest/data-prepper/pipelines/expression-syntax/) documentation.
 
+## Grok performance metadata
+
+When the `performance_metadata` option is set to true, the `grok` processor adds the following metadata keys to each event:
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:54]: Should "true" be in code font?

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 13]
```diff
@@ -46,8 +45,36 @@ processor:
...
+
+* `_total_grok_processing_time`: The total time, in milliseconds, that the `grok` processor takes to match the event. This is the sum of the processing time based on all of the `grok` processors which have the `performance_metadata` option enabled that ran on this event.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:56]:
```suggestion
* `_total_grok_processing_time`: The total amount of time, in milliseconds, that the `grok` processor takes to match the event. This is the sum of the processing time based on all of the `grok` processors that ran on the event and have the `performance_metadata` option enabled.
```

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 14]
```diff
@@ -46,8 +45,36 @@ processor:
...
+* `_total_grok_patterns_attempted`: The total number of `grok` pattern match attempts across all `grok` processors that have run on the event.
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:57]: We use "have run" here but "ran" in the preceding line. Should they match?

File: _data-prepper/pipelines/configuration/processors/grok.md [hunk 15]
```diff
@@ -46,8 +45,36 @@ processor:
...
+
+To include Grok performance metadata when the event is sent to the sink set inside the pipeline, use the `add_entries` processor tot describe the metadata you want to include, as shown in the following example:
```
Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:59]:
```suggestion
To include Grok performance metadata when the event is sent to the sink inside the pipeline, use the `add_entries` processor to describe the metadata you want to include, as shown in the following example:
```

Comment (natebower) [inline _data-prepper/pipelines/configuration/processors/grok.md:59]: I'm assuming that "set" isn't necessary here.

**Review by natebower — COMMENTED** [review]
@graytaylor0 @Naarcha-AWS Please see my comments and changes and let me know if you have any questions. Thanks!

File: _data-prepper/pipelines/configuration/processors/grok.md (hunk 12 above)
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:54]:
```suggestion
When the `performance_metadata` option is set to `true`, the `grok` processor adds the following metadata keys to each event:
```

File: _data-prepper/pipelines/configuration/processors/grok.md (hunk 14 above)
Comment (Naarcha-AWS) [inline _data-prepper/pipelines/configuration/processors/grok.md:57]:
```suggestion
* `_total_grok_patterns_attempted`: The total number of `grok` pattern match attempts across all `grok` processors that ran on the event.
```

**Pull request merged**